src/OpenTelemetry/Metrics/MetricPoint.cs

```diff
@@ -545,5 +545,47 @@ internal void TakeSnapshot(bool outputDelta)
         {
             throw new NotSupportedException($"{methodName} is not supported for this metric type.");
         }
+
+#pragma warning disable SA1204 // Static elements should appear before instance elements
+        private static int FindHistogramBucketIndexLinear(double[] bounds, double number)
+#pragma warning restore SA1204 // Static elements should appear before instance elements
+        {
+            int i;
+
+            for (i = 0; i < bounds.Length; i++)
+            {
+                if (number <= bounds[i])
+                {
+                    break;
+                }
+            }
+
+            return i;
+        }
+
+        private static int FindHistogramBucketIndexBinary(double[] bounds, double number)
```

Comment:
So `FindHistogramBucketIndexLinear` walks the array from beginning to end. What _should_ happen in that case is the JIT will elide (remove) the automatic bounds check on the array, which is something .NET always does to make sure you don't access memory you shouldn't. Part of its safety system, basically. `FindHistogramBucketIndexBinary` is accessing the array randomly from the JIT's perspective so we're _probably_ paying for bounds checks every access into the array, making the binary search much slower. Try this...

* Add a target to netstandard2.1. 

* Call the span BinarySearch:

```csharp
#if NETSTANDARD2_1_OR_GREATER
        private static int FindHistogramBucketIndexBinary(double[] bounds, double number)
        {
		    Span<double> buffer = new Span<double>(bounds);
            return buffer.BinarySearch(number);
        }
#else
        private static int FindHistogramBucketIndexBinary(double[] bounds, double number)
        {
            int left = 0;
            int right = bounds.Length - 1;

            while (left <= right)
            {
                int mid = (left + right) / 2;
                if (number == bounds[mid])
                {
                    return mid;
                }
                else if (number > bounds[mid])
                {
                    left = mid + 1;
                }
                else
                {
                    right = mid - 1;
                }
            }

            return right + 1;
        }
#endif
```

What does that do to the perf numbers? Might make it much faster on targets that have the span API available.

It looks like the [implementation](https://github.com/dotnet/runtime/blob/c37f6a26a31d3d7211017f7b9d1fe8540909b255/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.BinarySearch.cs#L40) is calling into `Unsafe.Add` to access the value in the array, probably to avoid the bounds check in the tight inner loop. Also has some micro-optimizations around `unit` to save some CPU ops.

Reply:
@CodeBlanch Do you know if `Array.BinarySearch` would be a good candidate here?

Reply:
Perhaps check https://github.com/prom-client-net/prom-client/blob/master/src/Prometheus.Client/HistogramImpl/HistogramHighBucketsStore.cs?

Reply:
@utpilla I went and checked it out. Lot's of overhead in the implementation of Array.BinarySearch figuring out the type. If it figures out the type is primitive, it calls into the [span API](https://github.com/dotnet/runtime/blob/dfa39de33fe7e0c883eb0c343ab4012c8d5351a8/src/libraries/System.Private.CoreLib/src/System/Array.cs#L766). So I would say, since we already know the type, better perf to just call directly into the span API and avoid the overhead.

Regarding adding the target, I know we would prefer to not do that but so much going into the new runtimes for perf we would really be limiting ourselves if we said we are going to stick with netstandard2.0 forever 😄

Reply:
After making the above code change and adding `netstandard2.1` to `OpenTelemetry.csproj` when I run the benchmark it hangs at the same spot. Might be a benchmarkdotnet bug... I'll leave this change out of this PR though

```text
// **************************
// Benchmark: HistogramBenchmarks.HistogramHotPath: DefaultJob [BoundCount=140]
// *** Execute ***
// Launch: 1 / 1
// Execute: dotnet "88f821f7-2814-4c83-a921-c127244d66cd.dll" --benchmarkName "Benchmarks.Metrics.HistogramBenchmarks.HistogramHotPath(BoundCount: 140)" --job "Default" --benchmarkId 15 in C:\Git\opentelemetry-dotnet\test\Benchmarks\bin\Release\net6.0\eeec8bcf-3904-4ae3-9a86-795630f3face\bin\Release\net6.0\88f821f7-2814-4c83-a921-c127244d66cd\bin\Release\net6.0
// BeforeAnythingElse

// Benchmark Process Environment Information:
// Runtime=.NET 6.0.5 (6.0.522.21309), X64 RyuJIT
// GC=Concurrent Workstation
// Job: DefaultJob

OverheadJitting  1: 1 op, 199400.00 ns, 199.4000 us/op
WorkloadJitting  1: 1 op, 6634400.00 ns, 6.6344 ms/op

OverheadJitting  2: 16 op, 239800.00 ns, 14.9875 us/op
```

Reply:
What does "hang" mean? Threads blocked waiting for something (without using CPU) or keep spinning (burning CPU)? If it is the latter, it might be due to a bug introduced by this PR.

Reply:
Might have been a code bug introduced. I will merging in the change @CodeBlanch emailed me about performing the binary search on a binary tree data structure. Marking as draft until this is added.